apps/renterd/lib/multipartUpload.ts

```diff
@@ -0,0 +1,293 @@
+    this.#onComplete = onComplete
+  }
+
+  async #sendNext() {
+    const activeConnections = Object.keys(this.#activeConnections).length
+
+    if (activeConnections >= this.#maxConcurrentChunks) {
+      return
+    }
+
+    if (!this.#pendingPartNumbers.length) {
+      if (!activeConnections) {
+        this.#complete()
+      }
+
+      return
+    }
+
+    const partNumber = this.#pendingPartNumbers.pop()
+    const partIndex = partNumber - 1
+    const chunkOffset = partIndex * this.#chunkSize
+    const chunk = this.#file.slice(chunkOffset, chunkOffset + this.#chunkSize)
+
+    // Callback to start another upload after the current one is added to the
+    // active connections.
+    // This will boot up the max concurrent uploads.
+    const tryStartingAnother = () => {
+      this.#sendNext()
+    }
+
+    try {
+      await this.#upload(partNumber, chunk, chunkOffset, tryStartingAnother)
+      // On successful upload, reset the delay
+      this.#resetDelay()
+    } catch (error) {
+      this.#pendingPartNumbers.push(partNumber)
+      await this.#waitToRetry()
+    }
+    // try again even after a part errors
+    this.#sendNext()
+  }
+
+  #resetDelay() {
+    this.#currentDelay = this.#initialDelay
+  }
+
+  async #waitToRetry() {
+    // Increase the delay for the next retry, capped at the maximum delay
+    const backoff = delay(this.#currentDelay)
+    this.#currentDelay = Math.min(this.#currentDelay * 2, this.#maxDelay)
+    await backoff
+  }
+
+  async #complete() {
+    try {
+      const payload = {
+        bucket: this.#bucket,
+        path: this.#path,
+        uploadID: this.#uploadId,
+        parts: this.#uploadedParts.sort((a, b) => a.partNumber - b.partNumber),
+      }
+      await this.#apiBusUploadComplete.post({
+        payload: payload,
+      })
+      this.#onComplete()
+    } catch (error) {
+      this.#onError(error)
+    }
+    this.#resolve()
+  }
+
+  #handleProgress(partNumber: number, event: ProgressEvent) {
+    this.#progressCache[partNumber] = event.loaded
+
+    const progressTotal = Object.keys(this.#progressCache)
+      .map(Number)
+      .reduce((acc, id) => (acc += this.#progressCache[id]), 0)
+
+    const sent = Math.min(progressTotal, this.#file.size)
+    const total = this.#file.size
+
+    const percentage = Math.round((sent / total) * 100)
+
+    this.#onProgress({
+      sent: sent,
+      total: total,
+      percentage: percentage,
+    })
+  }
+
+  async #upload(
+    partNumber: number,
+    chunk: Blob,
+    chunkOffset: number,
+    afterConnectionIsAdded: () => void
+  ): Promise<void> {
+    const controller = new AbortController()
+    this.#activeConnections[partNumber] = controller
+    afterConnectionIsAdded()
+    try {
+      const response = await this.#apiWorkerUploadChunk.put({
+        params: {
+          key: this.#path.slice(1),
+          bucket: this.#bucket,
+          uploadid: this.#uploadId,
+          offset: chunkOffset,
```

Comment:
might be a bit clearer to use the term "part" instead of "chunk" throughout the codebase.